probert/lvm.py

```diff
@@ -0,0 +1,241 @@
+
+    {'blkdevname': 'dm-0',
+     'blkdevs_used': 'sdbr,sdbq,sdbp',
+     'lv_name': 'lv_srv',
+     'subsystem': 'LVM',
+     'uuid': 'LVM-lyrZxQgOcgVSlj81LvyUnvq4DW3uLLrfJLI5ieYZR9a2fSOGBK03KM78',
+     'vg_name': 'storage_vg_242x'}
+    '''
+    _SEP = '='
+    fields = (
+        'subsystem,vg_name,lv_name,blkdevname,uuid,blkdevs_used'.split(','))
+    try:
+        output = subprocess.check_output(
+            ['sudo', 'dmsetup', 'info', devname, '-C', '-o',
+             ','.join(fields), '--noheading', '--separator', _SEP])
+    except subprocess.CalledProcessError as e:
+        log.error('Failed to probe dmsetup info:', e)
+        return None
+    values = output.decode('utf-8').strip().split(_SEP)
+    info = dict(zip(fields, values))
+    return info
+
+
+def lvmetad_running():
+    return os.path.exists(os.environ.get('LVM_LVMETAD_PIDFILE',
+                                         '/run/lvmetad.pid'))
+
+
+def lvm_scan(activate=True):
+    for cmd in [['pvscan'], ['vgscan', '--mknodes']]:
+        if lvmetad_running():
+            cmd.append('--cache')
+        try:
+            subprocess.run(cmd, stdout=subprocess.DEVNULL,
+                           stderr=subprocess.DEVNULL)
+        except subprocess.CalledProcessError as e:
+            log.error('Failed lvm_scan command %s: %s', cmd, e)
+
+
+def activate_volgroups():
+    """
+    Activate available volgroups and logical volumes within.
+    # found
+    % vgchange -ay
+      1 logical volume(s) in volume group "vg1sdd" now active
+
+    # none found (no output)
+    % vgchange -ay
+    """
+
+    # vgchange handles syncing with udev by default
+    # see man 8 vgchange and flag --noudevsync
+    result = subprocess.run(['vgchange', '--activate=y'], check=False,
+                            stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
+    if result.stdout:
+        log.info(result.stdout)
+
+
+def extract_lvm_partition(probe_data):
+    lv_id = "%s/%s" % (probe_data['DM_VG_NAME'], probe_data['DM_LV_NAME'])
+    return (
+        lv_id, {'fullname': lv_id,
+                'name': probe_data['DM_LV_NAME'],
+                'volgroup': probe_data['DM_VG_NAME'],
+                'size': "%sB" % read_sys_block_size(probe_data['DEVNAME'])})
+
+
+def extract_lvm_volgroup(vg_name, report_data):
+    """
+    [
+        {"vg_name":"vg0", "pv_name":"/dev/md0",
+         "pv_uuid":"p3oDow-dRHp-L8jq-t6gQ-67tv-B8B6-JWLKZP",
+         "vg_size":"21449670656B"},
+        {"vg_name":"vg0", "pv_name":"/dev/md1",
+         "pv_uuid":"pRR5Zn-c4a9-teVZ-TFaU-yDxf-FSDo-cORcEq",
+         "vg_size":"21449670656B"}
+    ]
+    """
+    devices = set()
+    for report in report_data:
+        if report['vg_name'] == vg_name:
+            size = report['vg_size']
+            devices.add(report.get('pv_name'))
+
+    return (vg_name, {'name': vg_name,
+                      'devices': list(devices),
+                      'size': size})
```

Comment:
I'm not quite sure how but I managed to hit a NameError on this size here (I may have been running as non-root). I guess the for loop didn't find anything. I'm not sure what should happen here but a NameError is not great :)

Reply:
Do you have the json data generated?  i'd really like to se the report_data  json.

It may be a linear VG, I've seen things where if the allocated space in a VG does not span one PV, then the second PV is effectively unallocated and could not have a size.

I'll refactor the loop to set size only if it's unset so we don't clobber the value if other members of the vg don't have a size.